js/dactylo.js

```diff
@@ -30,69 +32,96 @@ window.addEventListener('DOMContentLoaded', () => {
-    fetch(`dicts/${gDict.value}.json`)
+  // fetch MonkeyType words
+  const fetchWords = () => {
+    const words = gDict.value.split(',')[1];
+    return fetch(`data/dicts/${words}.json`)
       .then(response => response.json())
       .then(data => {
-        gDictionary = data.words;
-        setLessonLevel();
+        gDictionary.words = data.words;
       });
   };
 
-  gLayout.addEventListener('change', setLayout);
-  setLayout();
+  // fetch a kalamine keyboard layout
+  const fetchLayout = () => {
+    return fetch(`data/layouts/${gLayout.value}.json`)
+      .then(response => response.json())
+      .then(layout => {
+        gKeyboard.setKeyboardLayout(layout.keymap, layout.deadkeys, gGeometry.value);
+        gKeyboard.theme = 'hints';
+        gKeyLayout = layout;
+      });
+  };
 
-  gDict.addEventListener('change', setDict);
-  setDict();
+  gLayout.addEventListener('change', () => {
+    fetchLayout().then(setLessonLevel);
+  });
+
+  gDict.addEventListener('change', () => {
+    Promise.all([fetchNgrams(), fetchWords()]).then(setLessonLevel);
+  });
 
   gGeometry.addEventListener('change', event => {
     gKeyboard.geometry = gGeometry.value;
   });
 
   gKeyList.addEventListener('click', event => {
     if (event.target.nodeName.toLowerCase() == 'kbd') {
-      setLessonLevel(event.target.dataset.level);
+      gLessonLevel = event.target.dataset.level;
+      setLessonLevel();
     }
   });
 
-  const setLessonLevel = (level = gLessonLevel) => {
-    if (!gKeyLayout || !gDictionary) {
-      return;
-    }
-
-    const keys = g30Keys.slice(0, level);
-    const rawLetters = keys.flatMap(key => gKeyLayout.keymap[key]);
+  const setLessonLevel = () => {
+    const keys = ALL_30_KEYS.slice(0, gLessonLevel);
+    // const rawLetters = keys.flatMap(key => gKeyLayout.keymap[key]);
+    const rawLetters = keys.map(key => gKeyLayout.keymap[key][0]);
 
     const odk = gKeyLayout.deadkeys['**'];
     const has1dk = keys.some(key => gKeyLayout.keymap[key].indexOf('**') >= 0);
     const deadkeyLetters = !has1dk ? [] :
       rawLetters
         .filter(letter => letter in odk)
         .map(letter => odk[letter]);
+
     const lessonLetters = rawLetters.concat(deadkeyLetters);
+    const lessonFilter = word =>
+      Array.from(word).every(letter => lessonLetters.indexOf(letter) >= 0);
 
-    gLessonLevel = level;
-    gLessonWords = gDictionary.filter(word =>
-      Array.from(word).every(letter => lessonLetters.indexOf(letter) >= 0));
+    // gLessonLevel = level;
+    gLessonWords = gDictionary.words.filter(lessonFilter);
+    if (gLessonWords.length < MIN_WORD_WOUNT) {
+      gLessonWords = gLessonWords.concat(gDictionary.trigrams.filter(lessonFilter));
+    }
+    if (gLessonWords.length < MIN_WORD_WOUNT) {
+      gLessonWords = gLessonWords.concat(gDictionary.bigrams.filter(lessonFilter));
+    }
+    if (gLessonWords.length < MIN_WORD_WOUNT) {
+      gLessonWords = gLessonWords.concat(rawLetters);
+    }
```

Comment:
Oh hey, you’re learning this imperative + functional pattern pretty fast !

Reply:
This is by far the worst part of the whole file. :-)
it could be factorized with a while loop and a break statement.